docs/src/main/asciidoc/cdi.adoc

```diff
@@ -0,0 +1,361 @@
+No scope annotation is declared and so it's defaulted to `@Dependent`.
+<2> The container will delegate to this method when creating the bean instance.
+<3> `PiProducer#producePi()` is invoked and the result is set to `PiConsumer#pi` when the `PiConsumer` instance is created.
+
+NOTE: One of the limitations of producers is that interceptors are not applied to the result of a producer method/field.
+
+Producer fields are very similar to producer methods except that the field value is read when obtaining a bean instance.
+
+TIP: There's more about producers. You can declare qualifiers, inject dependencies into the producer methods parameters, etc. You can read more about producers for example in the https://docs.jboss.org/weld/reference/latest/en-US/html/beanscdi.html#_producer_methods[Weld docs, window="_blank"].
+
+Synthetic beans are usually provided by extensions.
+A synthetic bean is not backed by a class/method/field.
+Application developers usually only consume this kind of beans.
+
+== _Q: OK injection looks cool. What other services are provided?_
+
+=== Lifecycle Callbacks
+
+A bean class may declare lifecycle `@PostConstruct` and `@PreDestroy` callbacks:
+
+.Lifecycle Callbacks Example
+[source,java]
+----
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+
+@ApplicationScoped
+public class Translator {
+
+    @PostConstruct <1>
+    void init() {
+       // ...
+    }
+    
+    @PreDestroy <2>
+    void destroy() {
+      // ...
+    }
+}
+----
+<1> This callback is invoked before the bean instance is put into service. It is safe to perform some initialization here.
+<2> This callback is invoked before the bean instance is destroyed. It is safe to perform some cleanup tasks here.
+
+TIP: It's a good practice to keep the logic in the callbacks "without side effects", i.e. you should avoid calling other beans inside the callbacks.
+
+=== Interceptors
+
+Interceptors are used to separate cross-cutting concerns from business logic. 
+There is a separate specification - Java Interceptors - that defines the basic programming model and semantics.
+
+.Simple Interceptor Example
+[source,java]
+----
+import javax.interceptor.Interceptor;
+import javax.annotation.Priority;
+
+@Logged <1>
+@Priority(2020) <2>
+@Interceptor <3>
+public class LoggingInterceptor {
+
+   @Inject <4>
+   Logger logger;
+
+   @AroundInvoke <5>
+   Objec logInvocation(InvocationContext context) {
+      // ...log before
+      Objec ret = context.proceed(); <6>
+      // ...log after
+      return ret;
+   }
+    
+}
+----
+<1> This is an interceptor binding annotation that is used to bind our interceptor to a bean. Simply annotate a bean class with `@Logged`.
+<2> `Priority` enables the interceptor and affects the the interceptor ordering. 
+<3> Marks an interceptor component.
+<4> An interceptor instance may be the target of dependency injection.
+<5> `AroundInvoke` denotes a method that interposes on business methods.
```

Comment:
Maybe list the other interceptor pointcut ? (BeforeInvoke, AfterInvoke)

Reply:
I wouldn't complicate things here... BTW there is nothing like `BeforeInvoke` and `AfterInvoke` in the interceptors spec. There are `@AroundConstruct` and `@PostConstruct` and `@PreDestroy` interceptors...